Framework/PythonInterface/plugins/algorithms/RefineSatellitePeaks.py

```diff
@@ -0,0 +1,181 @@
+from mantid.kernel import *
+from mantid.api import *
+from mantid.simpleapi import *
+
+import numpy as np
+from scipy.spatial import KDTree
+from scipy.cluster.vq import kmeans2
+import scipy.cluster.hierarchy as hcluster
+
+
+class RefineSatellitePeaks(DataProcessorAlgorithm):
+
+    def category(self):
+        return 'Crystal\\Peaks'
+
+    def PyInit(self):
+        self.declareProperty(IPeaksWorkspaceProperty(name="MainPeaks",
+                                                     defaultValue="",
+                                                     direction=Direction.Input),
+                             doc="Main integer HKL peaks. Q vectors will be calculated relative to these peaks.")
+
+        self.declareProperty(IPeaksWorkspaceProperty(name="SatellitePeaks",
+                                                     defaultValue="",
+                                                     direction=Direction.Input),
+                             doc="Positions of seed satellite peaks. These will be used to define the q vectors for each satellite.")
+
+        self.declareProperty(WorkspaceProperty(name="MDWorkspace",
+                                               defaultValue="",
+                                               direction=Direction.Input),
+                             doc="Workspace to search for satellites peak in.")
+
+        self.declareProperty(IPeaksWorkspaceProperty(name="OutputWorkspace",
+                                                     defaultValue="",
+                                                     direction=Direction.Output),
+                             doc="All found satellite peaks. These will be given with satellite coordinates.")
+
+        self.declareProperty('NumOfQs', -1, direction=Direction.Input,
+                             doc="The number of satellite peaks to look for. If this option is not set to the default then all the \
+                             provided satellites will be grouped into exactly this number of q vectors")
+        self.declareProperty('ClusterThreshold', 1.5, direction=Direction.Input,
+                             doc="Threshold for automaticallty deciding on the number of q vectors to use. If NumOfQs found is set then this \
+                             is property is ignored.")
+        self.declareProperty('PeakRadius', 0.1, direction=Direction.Input,
+                             doc="The peak radius used to integrate the satellite peaks. This is passed direclty to IntegratePeaksMD")
+        self.declareProperty('BackgroundInnerRadius', 0.1, direction=Direction.Input,
+                             doc="The inner background radius used to integrate the satellite peaks. This is passed direclty to \
+                             IntegratePeaksMD")
```

Comment:
FYI, I think it's worth stating that the ESS will be trying to move away from `MDWorkspaces` as part of algorithm inputs. As this fits the needs of ISIS/SNS, that shouldn't be a reason to do anything differently.

Reply:
Thanks for the info. For them it's easiest to take an MD workspace as they want to manually inspect the diffraction pattern in the slice viewer and find their satellite peaks. As they are already creating the MD workspace for visualisation then I think having it as input is ok.